### ISSUE TITLE
Publishing images on the Public ECR registry

### ISSUE THREAD

**Issue by 094459** [valkey-io/valkey-container#17]
Amazon ECR has a public repository (https://gallery.ecr.aws/) where many developers set as their preferred container registry. There is currently no valkey container image.

**Comment by emsu-io-suefke** [comment]
This would indeed be helpful

**Comment by dnrce** [comment]
Related: valkey-io/valkey#1605

**Comment by madolson** [comment]
@roshkhatri Do you know how much work this would be to do?

**Comment by roshkhatri** [comment]
@madolson Should not take long, probably a day for each should be more than enough

**Comment by 094459** [comment]
We need to be careful - a one time image push is probably not going to cut it. As new versions, updates, patches, etc are produced, these need to be pushed to ECR. So this is more an ongoing activity that needs to be managed.

**Comment by roshkhatri** [comment]
Yes, we would have a workflow step to do this here: https://github.com/valkey-io/valkey-container/blob/mainline/.github/workflows/ci.yml.

This would basically enable us to build and push for every release.

**Comment by madolson** [comment]
Yeah, we shouldn't do it as a one off, we should build the automation so each time we release the main container we also release one on ECR.

**Comment by bjw-s** [comment]
I have submitted a PR that should allow pushing the images to GitHub Container Registry (as requested in #9). If GHCR ends up not being a desired target registry hopefully the PR can serve as a base for adding a different alternative registry.

**Comment by 094459** [comment]
One thing I could not see from the GitHub automation above was whether we are making sure that both AMD64 and ARM images are being built. Given that a significant number of folks will want to run this on Arm, need to make sure that the built images support both arcitectures.

**Comment by zc-devs** [comment]
JFYI, the images distribution across the registries could be done by copying images from the "main" registry to the "replicas" by [skopeo](https://github.com/containers/skopeo). There are the credentials helpers for [ECR](https://github.com/awslabs/amazon-ecr-credential-helper), [GCR](https://github.com/GoogleCloudPlatform/docker-credential-gcr) and [ACR](https://github.com/chrismellard/docker-credential-acr-env).

**Comment by bjw-s** [comment]
> One thing I could not see from the GitHub automation above was whether we are making sure that both AMD64 and ARM images are being built. Given that a significant number of folks will want to run this on Arm, need to make sure that the built images support both arcitectures.

In the PR, the same platforms get built for both Docker Hub and GHCR. The main difference is that the images are now being tagged and pushed to both registries. (perhaps it's better to have any follow-up discussion specific to my PR in the PR comments directly to avoid cluttering this issue)

**Comment by roshkhatri** [comment]
I think I am pretty convinced for having multiple distribution but just had a few questions.
1. Why does someone have a preferred image distribution(container registries)?
2. How does it help the projects to pull from preferred directories?

**Comment by joryirving** [comment]
> Why does someone have a preferred image distribution(container registries)?
How does it help the projects to pull from preferred directories?

Docker rate limits are a large one. If I'm say, bootstrapping a kubernetes cluster, or running something in concurrency on a GH runner, you can hit rate limits really quickly.

Alternate registries (like ghcr, public ECR), have no/higher pull limits. I personally always try to pull from GHCR first, public ECR second, and dockerhub if there's no alternatives, both in my homelab, and professionally at work.

**Comment by zc-devs** [comment]
> Why?

Apart from rate limiting, some registries

1. may deny access by IP
2. may become paid
3. may not work on IPv6
4. just might be not available at the moment

**Comment by zc-devs** [comment]
[Prepare now: Docker Hub rate limits will impact GitLab CI/CD](https://about.gitlab.com/blog/2025/03/24/prepare-now-docker-hub-rate-limits-will-impact-gitlab-ci-cd/)

> Unauthenticated users: 10 pulls per hour per IPv4 address or IPv6 /64 subnet

**Comment by roshkhatri** [comment]
Hey, Thank you for the patience and your feedback from the community.

We are now publishing the images on GHCR: https://github.com/valkey-io/valkey-container/pkgs/container/valkey

**Comment by mcg** [comment]
Is work still being done to push images to gallery.ecr.aws?

**Comment by roshkhatri** [comment]
Yes, we will be working on this soon.

**Comment by roshkhatri** [comment]
@valkey-io/core-team 
Can you take a look again, the community would want valkey image on Amazon ECR as well.

**Comment by madolson** [comment]
Yeah, you can go ahead and add support for Amazon ECR.

**Comment by roshkhatri** [comment]
Great, Thank you so much

**Comment by hanxizh9910** [comment]
I am looking at this issue

**Comment by hanxizh9910** [comment]
Hi, the Valkey container image is now available on the public ECR repository here: https://gallery.ecr.aws/valkey/valkey

**Comment by roshkhatri** [comment]
Thank you so much @hanxizh9910, I will close this issue as we are now live on:
**Dockerhub:** https://hub.docker.com/r/valkey/valkey
**Github Container Registry:** https://github.com/valkey-io/valkey-container/pkgs/container/valkey
**Amazon public ECR**: https://gallery.ecr.aws/valkey/valkey
